### PR TITLE
3d warmup

### PR DESCRIPTION
**Pull request by iboiko-habana**
3d warmup

### CODE REVIEW

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 1]
```diff
@@ -512,14 +523,42 @@ def trim_attn_metadata(self, metadata: AttentionMetadata) -> object:
...
+                phase_type = PhaseType.DECODE
+            else:
+                raise ValueError(
+                    "Unrecognized pass type, likely due to malformed "
+                    "attention metadata")
+        else:
+            return 'prompt' if is_prompt else 'decode'
+        return phase_type.value
+
+    def _check_config(self, batch_size, seq_len, ctx, attn_metadata,
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_enc_dec_model_runner.py:545]: why can't we get ctx from metadata (as we do with blocks)?

File: vllm/worker/hpu_model_runner.py [hunk 2]
```diff
@@ -1274,9 +1280,14 @@ def _prepare_prompt(
             target_query_len = max(query_lens)
         real_num_seqs = len(query_lens)
 
-        max_prompt_len = max(
-            self.bucketing_ctx.get_padded_prompt_seq_len(target_query_len),
-            self.block_size)
+        if sum(context_lens) != 0:
+            max_prompt_len = max(
+                self.bucketing_ctx.get_padded_prefix_prefill_seq_len(
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_model_runner.py:1285]: why we can't use the same function? Why we need to have separate padded_prefix_prefill_seq_len? In this context seq_len stays the same as afaik it is the number of query_tokens (in this context).

File: vllm/worker/hpu_model_runner.py [hunk 3]
```diff
@@ -2002,17 +2013,22 @@ def create_dummy_seq_group_metadata(self,
...
             sampling_params = SamplingParams(temperature=temperature)
             num_blocks = math.ceil(seq_len / self.block_size)
         seq_len = max(seq_len, 1)
+        computed_block_nums = None
         if is_prompt:
             input_len = seq_len
             output_len = 0
             block_tables = None
+            if ctx:
+                block_tables = {group_id: [_PAD_BLOCK_ID] * ctx * 128}
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_model_runner.py:2030]: I don't think this is right...
First of all, we shouldn't use 128. I think we already have block_size in a variable. self.block_size ?
Secondly, block_table contains whole blocks. So if we have ctx of N, this means we need N blocks, so there's no need to multiply it by block size.

File: vllm/worker/hpu_model_runner.py [hunk 4]
```diff
@@ -2229,21 +2252,30 @@ def warmup_graphs(self,
...
         captured_all = True
         warmed_random_sampler_bs: Set[int] = set()
-        for idx, (batch_size, seq_len) in enumerate(buckets):
+        for idx, (batch_size, seq_len, ctx) in enumerate(buckets):
             # Graph memory usage is proportional to seq dimension in a batch
-            batch_seq = batch_size * seq_len if is_prompt else batch_size
+            phase = f'Graph/{self._phase(is_prompt, ctx)}'
+            if is_prompt:
+                if ctx:
+                    batch_seq = batch_size * seq_len * ctx
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_model_runner.py:2260]: doesn't seq_len already contain ctx?

File: vllm/worker/hpu_model_runner.py [hunk 5]
```diff
@@ -2329,6 +2376,9 @@ def warmup_model(self, kv_caches: List[torch.Tensor]) -> None:
         ) if can_use_compile_only_mode else contextlib.nullcontext():
             self.warmup_all_buckets(self.bucketing_ctx.prompt_buckets, True,
                                     kv_caches)
+            if self.use_prefix_caching:
+                self.warmup_all_buckets(
+                    self.bucketing_ctx.prefix_prefill_buckets, True, kv_caches)
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_model_runner.py:2381]: why do we need a separate run for prefix_prefill? Can't we warm them up during regular prefill warmup?

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 6]
```diff
@@ -512,14 +523,42 @@ def trim_attn_metadata(self, metadata: AttentionMetadata) -> object:
...
+                      warmup_mode):
+        cfg: Optional[tuple] = None
```
[the remark this answers is not in the document]
Comment (adobrzyn) [reply vllm/worker/hpu_enc_dec_model_runner.py:547]: According to mypy - yes, it's mostly for style, i didn't want to add ignore for mypy

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 7]
```diff
@@ -512,14 +523,42 @@ def trim_attn_metadata(self, metadata: AttentionMetadata) -> object:
         ])
         return attention_metadata
 
-    def _check_config(self, batch_size, seq_len, is_prompt, warmup_mode):
-        cfg = (batch_size, seq_len, is_prompt)
+    def _phase(self, is_prompt, ctx):
+        if self.use_prefix_caching:
```
[the remark this answers is not in the document]
Comment (adobrzyn) [reply vllm/worker/hpu_enc_dec_model_runner.py:527]: I will remove it, the same as I did with hpu_model_runner - let's use prompt/decode only

File: vllm/worker/hpu_model_runner.py (hunk 4 above)
Comment (adobrzyn) [reply vllm/worker/hpu_model_runner.py]: I changed variable names to query so that seq_len = query len + ctx * blocks size. It will help differentiate those two things - here seq len was used as query len before

File: vllm/worker/hpu_model_runner.py (hunk 5 above)
Comment (adobrzyn) [reply vllm/worker/hpu_model_runner.py]: It was moved to one - thanks for this advice :)

File: vllm/v1/worker/hpu_model_runner.py [hunk 8]
```diff
@@ -1469,29 +1473,66 @@ def _execute_model_generic(self,
         batch_size = token_ids.size(0)
         seq_len = self._seq_len(attn_metadata)
         num_blocks = self._num_blocks(attn_metadata)
-
+        #print("before", seq_len, num_blocks)
```
[the remark this answers is not in the document]
Comment (adobrzyn) [reply vllm/v1/worker/hpu_model_runner.py]: Removed, thx

File: vllm/v1/worker/hpu_model_runner.py [hunk 9]
```diff
@@ -616,7 +610,7 @@ def __init__(
             self.bucketing_ctx = HPUBucketingContext(
                 self.max_num_seqs, self.max_prefill_batch_size,
                 self.block_size, self.max_num_batched_tokens,
-                self.use_merged_prefill, self.max_model_len)
+                self.use_merged_prefill, self.max_model_len, prefix_caching=False)
```
Comment (madamczyk-intel) [inline vllm/v1/worker/hpu_model_runner.py:613]: why we always pass false? Shouldn't we take it from configuration?

File: vllm/v1/worker/hpu_model_runner.py [hunk 10]
```diff
@@ -2017,56 +2053,44 @@ def log_warmup(self, phase, i, max_i, batch_size, seq_len):
...
             # Graph memory usage is proportional to seq dimension in a batch
-            batch_seq = batch_size * seq_len if is_prompt else batch_size
-            mem_estimate = batch_seq / total_batch_seq * total_mem
-            if mem_estimate >= available_mem:
-                captured_all = False
-                continue
-            graphed_bucket = (batch_size, seq_len, is_prompt)
+            phase = f"Graph/{'prompt' if is_prompt else 'decode'}"
+            #batch_seq = batch_size * seq_len if is_prompt else batch_size
+            if is_prompt:
```
Comment (madamczyk-intel) [inline vllm/v1/worker/hpu_model_runner.py:2075]: is `batch_seq` used?

File: vllm/v1/worker/hpu_model_runner.py [hunk 11]
```diff
@@ -2017,56 +2053,44 @@ def log_warmup(self, phase, i, max_i, batch_size, seq_len):
...
+            self.log_warmup(phase, idx, num_candidates, batch_size, seq_len,
+                            num_blocks)
             with HabanaMemoryProfiler() as mem_prof:
-                self.warmup_scenario(batch_size, seq_len, is_prompt, kv_caches)
+                self.warmup_scenario(batch_size, seq_len, num_blocks,
+                                     is_prompt, kv_caches)
             #TODO(kzawora): align_workers
             used_mem = mem_prof.consumed_device_memory
-            available_mem -= used_mem
             total_mem += used_mem
```
Comment (madamczyk-intel) [inline vllm/v1/worker/hpu_model_runner.py:2094]: is it still used?

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 12]
```diff
@@ -340,23 +340,26 @@ def profile_run(self) -> None:
         max_seq_len = min(self.max_num_batched_tokens // max_batch_size,
                           max_seq_len)
 
-        self.warmup_scenario(max_batch_size, max_seq_len, True, kv_caches,
+        self.warmup_scenario(max_batch_size, max_seq_len, 0, True, kv_caches,
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_enc_dec_model_runner.py:343]: having both seq_len and blocks for decode is confusing as we can pass (bs, 1, num_blocks) or (bs, num_blocks * block_size, 0)

File: vllm/worker/hpu_model_runner.py [hunk 13]
```diff
@@ -1036,12 +1032,12 @@ def get_model(self) -> torch.nn.Module:
...
         return self.model
 
-    def _use_graphs(self, batch_size, seq_len, is_prompt):
+    def _use_graphs(self, batch_size, seq_len, ctx, is_prompt):
         if self.enforce_eager:
             return False
         if self.skip_warmup:
             return True
-        return (batch_size, seq_len, is_prompt) in self.graphed_buckets
+        return (batch_size, seq_len, ctx, is_prompt) in self.graphed_buckets
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_model_runner.py:1040]: We're only warming up graphs, so this function doesn't make much sense now. It could be simplified to:
```
return not self.enforce_eager
```

File: vllm/v1/worker/hpu_model_runner.py (hunk 10 above)
Comment (adobrzyn) [reply vllm/v1/worker/hpu_model_runner.py:2075]: But this comment should disappear

File: vllm/v1/worker/hpu_model_runner.py (hunk 11 above)
Comment (adobrzyn) [reply vllm/v1/worker/hpu_model_runner.py:2094]: It's for logs

File: vllm/worker/hpu_model_runner.py (hunk 13 above)
Comment (adobrzyn) [reply vllm/worker/hpu_model_runner.py]: Fixed

File: vllm/v1/worker/hpu_model_runner.py [hunk 14]
```diff
@@ -1662,20 +1694,27 @@ def execute_model(
...
-                    logits=logits_device, sampling_metadata=sampling_metadata)
+
+                with self.profiler.record_event('internal',
+                                                "_prepare_sampling"):
+                    sampling_metadata = self._prepare_sampling(
+                        batch_changed, req_id, pad_to=logits_device.shape[0])
+                    sampler_output = self.sampler(
+                        logits=logits_device,
+                        sampling_metadata=sampling_metadata)
+
```
[the remark this answers is not in the document]
Comment (adobrzyn) [reply vllm/v1/worker/hpu_model_runner.py]: It was removed, will not be added in this PR

File: vllm/v1/worker/hpu_model_runner.py [hunk 15]
```diff
@@ -1433,16 +1426,19 @@ def _num_blocks(self, attn_metadata):
...
         is_prompt = attn_metadata.is_prompt
-        is_prefix_cached = is_prompt and attn_metadata.block_list is not None
+        '''is_prefix_cached = is_prompt and attn_metadata.block_list is not None
         if is_prompt and is_prefix_cached:
             phase_type = PhaseType.PREFIX_PREFILL
         elif is_prompt and not is_prefix_cached:
             phase_type = PhaseType.PREFILL
         elif not is_prompt:
-            phase_type = PhaseType.DECODE
+            phase_type = PhaseType.DECODE'''
```
[the remark this answers is not in the document]
Comment (adobrzyn) [reply vllm/v1/worker/hpu_model_runner.py]: Removed

**Review by michalkuligowski — COMMENTED** [review]
Please resolve conflicts

File: vllm/worker/hpu_model_runner.py [hunk 16]
```diff
@@ -2777,9 +2737,11 @@ def _warmup_multimodal_graph(self,
 
     def log_graph_warmup_summary(self, buckets, is_prompt, total_mem):
         num_candidates = len(buckets)
-        phase = f'Graph/{"Prompt" if is_prompt else "Decode"}'
-        graphed = list(c[:2] for c in self.graphed_buckets
-                       if c[2] == is_prompt)
+        phase = 'prompt' if is_prompt else 'decode'
+        # TODO - verify
```
Comment (madamczyk-intel) [inline vllm/worker/hpu_model_runner.py:2741]: stale comment

**Review by madamczyk-intel — CHANGES_REQUESTED** [review]

File: vllm/v1/worker/hpu_model_runner.py [hunk 17]
```diff
@@ -2212,6 +2217,9 @@ def warmup_model(self) -> None:
         if prompt_profile_cfg or decode_profile_cfg:
             self._generate_profiling(prompt_profile_cfg, decode_profile_cfg)
             raise AssertionError("Finished profiling")
+        if self.skip_warmup:
```
[the remark this answers is not in the document]
Comment (adobrzyn) [reply vllm/v1/worker/hpu_model_runner.py:2220]: Oki, I will remove that

File: vllm/worker/hpu_enc_dec_model_runner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (madamczyk-intel) [reply vllm/worker/hpu_enc_dec_model_runner.py]: ok, let's fix this in future PRs
